assertj-core/src/main/java/org/assertj/core/api/recursive/comparison/RecursiveComparisonConfiguration.java

```diff
@@ -792,6 +835,19 @@ private boolean matchesAnIgnoredFieldType(DualValue dualValue) {
     return false;
   }
 
+  private boolean shouldBeComparedBasedOnFieldValue(DualValue dualValue) {
+    if (parentDualValuePathsComparedFieldsOfTypes.stream().anyMatch(path -> dualValue.fieldLocation.getPathToUseInRules().startsWith(path))) return matchFiledComparedByTypeAndStorePath(dualValue);
+    if (dualValue.actual != null && comparedFieldsOfTypes.contains(dualValue.actual.getClass())) return matchFiledComparedByTypeAndStorePath(dualValue);
+    if (dualValue.expected != null && comparedFieldsOfTypes.contains(dualValue.expected.getClass())) return matchFiledComparedByTypeAndStorePath(dualValue);
```

Comment:
Added javadoc on how is compared type evaluated.

We don't evaluate `dualValue.expected` always now, we'll get to the `dualValue.expected`'s `if` statement only if `dualValue.actual` is `null`

However we can remove `matchFiledComparedByTypeAndStorePath(DualValue dualValue)` method, which only adds `FieldLocation` to parent compared fields set and always returns `true` and modify `shouldBeComparedBasedOnFieldValue(DualValue dualValue)` to something like:

```java
  private boolean shouldBeComparedBasedOnFieldValue(DualValue dualValue) {
    if (fieldLocationOfComparedFieldsOfTypes.stream().anyMatch(dualValue.fieldLocation::hasParent)) {
      fieldLocationOfComparedFieldsOfTypes.add(dualValue.fieldLocation);
      return true;
    }
    if (dualValue.actual != null && comparedFieldsOfTypes.contains(dualValue.actual.getClass())) {
      fieldLocationOfComparedFieldsOfTypes.add(dualValue.fieldLocation);
      return true;
    }
    if (dualValue.expected != null && comparedFieldsOfTypes.contains(dualValue.expected.getClass())) {
      fieldLocationOfComparedFieldsOfTypes.add(dualValue.fieldLocation);
      return true;
    }

    return false;
  }
```

So that might be more readable.

To be honest I don't have any strong argument for any approach here so we can just pick one - more obvious or less repetitiveness.